### PR TITLE
[BugFix] Fix materialized view rewrite when there is no partition to compensate

### PR DESCRIPTION
**Pull request by LiShuMing**
- If there is no partition selected in the  query, we can still use the mv to rewrite.
- add global rewrite log in `trace rewrite`

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 59 / 62 (95.16%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|12|14|85.71%|[200, 239]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|21|22|95.45%|[919]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 58 / 61 (95.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|12|14|85.71%|[200, 239]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|20|21|95.24%|[919]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerTraceUtil.java [hunk 1]
```diff
@@ -188,6 +195,13 @@ public static void logMVRewrite(OptimizerContext optimizerContext, Rule rule,
                     rule.type().name(),
                     String.format(format, object));
         }
+
+        // Trace log if needed.
+        PlannerProfile.LogTracer tracer = PlannerProfile.getLogTracer("REWRITE GLOBAL");
+        if (tracer != null) {
+            FormattingTuple ft = MessageFormatter.arrayFormat(format, object);
+            tracer.log(String.format("[%s] %s",   rule.type().name(), ft.getMessage()));
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerTraceUtil.java:203]:
```suggestion
            tracer.log(String.format("[%s] %s", rule.type().name(), ft.getMessage()));
```

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 38 / 50 (76.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|0|3|00.00%|[116, 123, 166]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|1|2|50.00%|[927]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|5|9|55.56%|[200, 201, 202, 203]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|9|11|81.82%|[947, 965]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|15|17|88.24%|[200, 240]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 37 / 48 (77.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|0|3|00.00%|[116, 123, 166]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|5|9|55.56%|[200, 201, 202, 203]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|9|11|81.82%|[947, 965]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|15|17|88.24%|[200, 240]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**